src/irr/asset/CGLSLBRDFBuiltinIncludeLoader.h

```diff
@@ -97,22 +113,82 @@ mat2x4 irr_glsl_applyScreenSpaceChainRule4D3(in mat3x4 dFdG, in mat2x3 dGdScreen
    return dFdG*dGdScreen;
 }
 
+//TODO move to different glsl header
+vec2 irr_glsl_concentricMapping(in vec2 _u)
+{
+    //map [0;1]^2 to [-1;1]^2
+    vec2 u = 2.0*_u - 1.0;
+    
+    vec2 p;
+    if (u==vec2(0.0))
+        p = vec2(0.0);
+    else
+    {
+        float r;
+        float theta;
+        if (abs(u.x)>abs(u.y)) {
+            r = u.x;
+            theta = 0.25*irr_glsl_PI * (u.y/u.x);
+        } else {
+            r = u.y;
+            theta = 0.5*irr_glsl_PI - 0.25*irr_glsl_PI*(u.x/u.y);
+        }
+        p = r*vec2(cos(theta),sin(theta));
+    }
+
+    return p;
+}
+
+//TODO move this to different glsl header
+mat2x3 irr_glsl_frisvad(in vec3 n)
+{
+	const float a = 1.0/(1.0 + n.z);
+	const float b = -n.x*n.y*a;
+	return (n.z<-0.9999999) ? mat2x3(vec3(0.0,-1.0,0.0),vec3(-1.0,0.0,0.0)):mat2x3(vec3(1.0-n.x*n.x*a, b, -n.x),vec3(b, 1.0-n.y*n.y*a, -n.y));
+}
+
 // only in the fragment shader we have access to implicit derivatives
-irr_glsl_ViewSurfaceInteraction irr_glsl_calcFragmentShaderSurfaceInteraction(in vec3 _CamPos, in vec3 _SurfacePos, in vec3 _Normal)
+irr_glsl_IsotropicViewSurfaceInteraction irr_glsl_calcFragmentShaderSurfaceInteraction(in vec3 _CamPos, in vec3 _SurfacePos, in vec3 _Normal)
 {
-   irr_glsl_ViewSurfaceInteraction interaction;
+   irr_glsl_IsotropicViewSurfaceInteraction interaction;
    interaction.V.dir = _CamPos-_SurfacePos;
    interaction.V.dPosdScreen[0] = dFdx(_SurfacePos);
    interaction.V.dPosdScreen[1] = dFdy(_SurfacePos);
    interaction.N = _Normal;
+   float invlenV2 = inversesqrt(dot(interaction.V.dir,interaction.V.dir));
+   float invlenN2 = inversesqrt(dot(interaction.N,interaction.N));
+   interaction.V.dir *= invlenV2;
+   interaction.N *= invlenN2;
```

Comment:
you can flat out call `normalize`

BUT

I would like two separate functions for `irr_glsl_calcFragmentShaderSurfaceInteraction`, one that assumes that `N` is unit length (and does not attempt to normalize it) and one that does not.

This is why instead of removing the `irr_glsl_ViewSurfaceInteraction` struct and its methods, I'd rather introduce a new one.

Reply:
let me know if tackled in new implementation.

Reply:
?